Gems/ROS2/Code/Source/RobotImporter/Utils/RobotImporterUtils.cpp

```diff
@@ -90,132 +89,294 @@ namespace ROS2
+                for (uint64_t linkIndex{}; linkIndex < currentModel.LinkCount(); ++linkIndex)
+                {
+                    const sdf::Link* link = currentModel.LinkByIndex(linkIndex);
+                    if (link != nullptr)
+                    {
+                        m_linkVisitorCB(*link);
+                    }
+                }
+            }
+
+        public:
+            LinkVisitorCallback m_linkVisitorCB;
+            bool m_recurseModels{};
+        };
+
+        VisitLinksForNestedModels_fn VisitLinksForNestedModels{};
+        VisitLinksForNestedModels.m_linkVisitorCB = linkVisitorCB;
+        VisitLinksForNestedModels.m_recurseModels = visitNestedModelLinks;
+        VisitLinksForNestedModels(sdfModel);
+    }
+
+    void VisitJoints(const sdf::Model& sdfModel, const JointVisitorCallback& jointVisitorCB,
+        bool visitNestedModelJoints)
+    {
+        // Function object which can visit all joints of a model
+        // Optionally it supports recursing nested models to visit their joints as well
+        struct VisitJointsForNestedModels_fn
+        {
+            void operator()(const sdf::Model& model)
+            {
+                VisitJointsForModel(model);
+                if (m_recurseModels)
+                {
+                    for (uint64_t modelIndex{}; modelIndex < model.ModelCount(); ++modelIndex)
+                    {
+                        const sdf::Model* nestedModel =  model.ModelByIndex(modelIndex);
+                        if (nestedModel != nullptr)
+                        {
+                            VisitJointsForModel(*nestedModel);
+                        }
+                    }
+                }
+            }
+
+        private:
+            void VisitJointsForModel(const sdf::Model& currentModel)
+            {
+                for (uint64_t jointIndex{}; jointIndex < currentModel.JointCount(); ++jointIndex)
+                {
+                    const sdf::Joint* joint = currentModel.JointByIndex(jointIndex);
+                    if (joint != nullptr)
+                    {
+                        m_jointVisitorCB(*joint);
+                    }
+                }
+            }
+
+        public:
+            JointVisitorCallback m_jointVisitorCB;
+            bool m_recurseModels{};
+        };
+
+        VisitJointsForNestedModels_fn VisitJointsForNestedModels{};
+        VisitJointsForNestedModels.m_jointVisitorCB = jointVisitorCB;
+        VisitJointsForNestedModels.m_recurseModels = visitNestedModelJoints;
+        VisitJointsForNestedModels(sdfModel);
+    }
+
+    AZStd::unordered_map<AZStd::string, const sdf::Link*> GetAllLinks(const sdf::Model& sdfModel,
+        bool gatherNestedModelLinks)
+    {
+        using LinkMap = AZStd::unordered_map<AZStd::string, const sdf::Link*>;
+        LinkMap links;
+        auto GatherLinks = [&links](const sdf::Link& link)
+        {
+            AZStd::string_view linkName(link.Name().c_str(), link.Name().size());
```

Comment:
Actually string_view storing into a string_view is done purposely here.
The `AZStd::string_view` is referencing the `std::string` data long though to pass into the unordered_map::insert_or_assign function, where it would initialize a `AZStd::string` from the `AZStd::string_view`.

In this case, it has slightly better performance (Only 1 copy) than copying the string first and then moving that it into links map(1 copy + 1 move)

Reply:
Ah, You are right about the `sdf::Link::Name` function

It returns a `std::string` by [value](http://osrf-distributions.s3.amazonaws.com/sdformat/api/13.2.0/classsdf_1_1SDF__VERSION__NAMESPACE_1_1Joint.html#ad9b4e04ebe8303ff0f4b1d7d0900488b).

Where the `sdf::Joint::Name` function returns a [const lvalue reference](http://osrf-distributions.s3.amazonaws.com/sdformat/api/13.2.0/classsdf_1_1SDF__VERSION__NAMESPACE_1_1Joint.html#ad9b4e04ebe8303ff0f4b1d7d0900488b) to a std::string.

The inconsistency in the SDF API caught me out here.

I will fix this code.